azure-pipelines.yml

```diff
@@ -1,144 +1,227 @@
-    condition: and(succeeded(), ne(variables['Build.Reason'], 'PullRequest'))
-
-  - script: docker push pythondiscord/snekbox-base:latest
-    displayName: 'Push Base Image to Dockerhub'
-    condition: and(succeeded(), ne(variables['Build.Reason'], 'PullRequest'), eq(variables.BASE_CHANGED, 'true'))
-
-  - script: docker push pythondiscord/snekbox-venv:latest
-    displayName: 'Push Virtual Environment Image to Dockerhub'
-    condition: and(succeeded(), ne(variables['Build.Reason'], 'PullRequest'), or(eq(variables.BASE_CHANGED, 'true'), eq(variables.VENV_CHANGED, 'true')))
-
-  - script: docker push pythondiscord/snekbox:latest
-    displayName: 'Push Final Image to Dockerhub'
-    condition: and(succeeded(), ne(variables['Build.Reason'], 'PullRequest'))
+  - job: test
+    displayName: 'Lint & Test'
+
+    pool:
+      vmImage: 'ubuntu-16.04'
+
+    steps:
+      - task: ShellScript@2
+        displayName: 'Check If Images Need to Be Built'
+        name: check
+        inputs:
+          scriptPath: scripts/check_dockerfiles.sh
+
+      # Without a login the following Docker build tasks won't add image tags.
+      - task: Docker@2
+        displayName: 'Log into Docker Hub'
+        inputs:
+          command: login
+          containerRegistry: DockerHubV2
+
+      # The venv image depends on this image. Build it if it can't be pulled
+      # from Docker Hub, which will be the case if the base Dockerfile has had
+      # changes.
+      - task: Docker@2
+        displayName: 'Build Base Image'
+        condition: and(succeeded(), ne(variables['check.BASE_PULL'], True))
+        inputs:
+          command: build
+          repository: pythondiscord/snekbox-base
+          tags: latest
+          Dockerfile: docker/base.Dockerfile
+          buildContext: .
+
+      # The dev image is never pushed and therefore is always built.
+      - task: Docker@2
+        displayName: 'Build Development Image'
+        inputs:
+          command: build
+          repository: pythondiscord/snekbox-venv
+          tags: dev
+          Dockerfile: docker/venv.Dockerfile
+          buildContext: .
+          arguments: --build-arg DEV=1
+
+      # The linter and all tests run inside this container.
+      - script: |
+          docker run \
+            --tty \
+            --detach \
+            --name snekbox_test \
+            --privileged \
+            --network host \
+            --hostname pdsnk-dev \
+            -e PYTHONDONTWRITEBYTECODE=1 \
+            -e PIPENV_PIPFILE="/snekbox/Pipfile" \
+            -e ENV="${PWD}/scripts/.profile" \
+            --volume "${PWD}":"${PWD}" \
+            --workdir "${PWD}"\
+            --entrypoint /bin/ash \
+            pythondiscord/snekbox-venv:dev
+        displayName: 'Start Container'
+
+      - script: |
+          docker exec snekbox_test /bin/ash -c \
+            'pipenv run lint --format junit-xml --output-file test-lint.xml'
+        displayName: 'Run Linter'
```

Comment:
with `--format junit-xml` do you not remove the default output?

Reply:
Yes it does. I would like to have both but not sure how.

Reply:
The `--tee` option would make it print to stdout but it'd still be in the XML format. The only solution I can think of is to run flake8 twice with different options. Thoughts?

Reply:
I think it would be better to just leave the JUnitXML reporting out. I also like it, but I think that relying on Azure to properly display the lint output is a bit unintuitive.

Reply:
I don't think it's too taxing to run the linting twice but I do admit I never really cared much for the fancy output for linting in particular. It would create separate entries for everything IIRC.

Testing seems to still output to stdout so that's fine. What about the coverage report, which doesn't output to stdout? The output on Azure is quite nice. Should the plain old `coverage report` be ran too in CI? I don't think it's taxing since it just reads the already made `.coverage` file.